File: site/queens/solver.js
// Queens solver: candidate elimination with single, confinement and no-room rules.
function QUEENS_SOLVER(U) {
  const QPALETTE = [['lavender', '#BBA3E2'], ['blue', '#96BEFF'], ['orange', '#FFC992'], ['gray', '#DFDFDF'], ['green', '#B3DFA0'], ['taupe', '#B9B29E'], ['red', '#FF7B60'], ['yellow', '#E6F388'], ['pink', '#DFA0BF'], ['teal', '#95CBCF'], ['salmon', '#FAA889'], ['cyan', '#55EBE2'], ['magenta', '#FE93F1'], ['mint', '#91F5AD'], ['periwinkle', '#C9C9EE'], ['emerald', '#5BBA6F']];
  const joinNames = names => names.length <= 1 ? (names[0] || '') : names.slice(0, -1).join(', ') + ' and ' + names[names.length - 1];
  function* combos(arr, k, start = 0, acc = []) {
    if (acc.length === k) { yield acc.slice(); return; }
    for (let i = start; i <= arr.length - (k - acc.length); i++) { acc.push(arr[i]); yield* combos(arr, k, i + 1, acc); acc.pop(); }
  }
  function queensPuzzle(n, regions, colors, names) {
    const labels = [], regionCells = {};
    for (let r = 0; r < n; r++) for (let c = 0; c < n; c++) { const l = regions[r][c]; if (!(l in regionCells)) { labels.push(l); regionCells[l] = []; } regionCells[l].push([r, c]); }
    if (!names) { names = {}; labels.forEach(l => { const hit = QPALETTE.find(p => p[1] === colors[l]); names[l] = hit ? hit[0] : 'colour ' + l; }); }
    const units = [];
    for (let i = 0; i < n; i++) units.push(['row', i]);
    for (let i = 0; i < n; i++) units.push(['col', i]);
    for (const l of labels) units.push(['region', l]);
    const unitCells = u => u[0] === 'row' ? [...Array(n)].map((_, c) => [u[1], c]) : u[0] === 'col' ? [...Array(n)].map((_, r) => [r, u[1]]) : regionCells[u[1]];
    const unitName = u => u[0] === 'row' ? `row ${u[1] + 1}` : u[0] === 'col' ? `column ${u[1] + 1}` : `the ${names[u[1]]} region`;
    const cellUnits = ([r, c]) => [['row', r], ['col', c], ['region', regions[r][c]]];
    const neighbours = ([r, c]) => { const out = []; for (let dr = -1; dr <= 1; dr++) for (let dc = -1; dc <= 1; dc++) { if ((dr || dc) && r + dr >= 0 && r + dr < n && c + dc >= 0 && c + dc < n) out.push([r + dr, c + dc]); } return out; };
    return { n, regions, labels, colors, names, units, unitCells, unitName, cellUnits, neighbours, regionCells };
  }
  const sameUnit = (a, b) => a[0] === b[0] && a[1] === b[1];
  class QState {
    constructor(p) { this.p = p; this.cand = Array.from({ length: p.n }, () => Array(p.n).fill(true)); this.queens = []; }
    copy() { const s = new QState(this.p); s.cand = this.cand.map(r => r.slice()); s.queens = this.queens.slice(); return s; }
    candidates(u) { return this.p.unitCells(u).filter(([r, c]) => this.cand[r][c]); }
    hasQueen(u) { const cells = this.p.unitCells(u); return this.queens.some(q => cells.some(x => x[0] === q[0] && x[1] === q[1])); }
    openUnits() { return this.p.units.filter(u => !this.hasQueen(u)); }
    eliminate(cells) { const removed = []; for (const [r, c] of cells) if (this.cand[r][c]) { this.cand[r][c] = false; removed.push([r, c]); } return removed; }
    struckBy(cell) {
      const set = new Set();
      for (const u of this.p.cellUnits(cell)) for (const [r, c] of this.p.unitCells(u)) set.add(r * this.p.n + c);
      for (const [r, c] of this.p.neighbours(cell)) set.add(r * this.p.n + c);
      set.delete(cell[0] * this.p.n + cell[1]);
      return set;
    }
    place(cell) {
      this.queens.push(cell); this.cand[cell[0]][cell[1]] = false;
      const keys = [...this.struckBy(cell)].sort((a, b) => a - b);
      return this.eliminate(keys.map(k => [Math.floor(k / this.p.n), k % this.p.n]));
    }
    ok() { return this.openUnits().every(u => this.candidates(u).length > 0); }
  }
  function qSingle(st) {
    const p = st.p;
    for (const u of st.openUnits()) {
      const cands = st.candidates(u);
      if (!cands.length) return 'contradiction';
      if (cands.length === 1) { const cell = cands[0]; return { kind: 'queen', cells: [cell], reason: `${p.unitName(u)} has one cell left`, cause: p.unitCells(u).filter(x => x[0] !== cell[0] || x[1] !== cell[1]) }; }
    }
    return null;
  }
  function qConfine(st, k) {
    const p = st.p, open = st.openUnits(), openRegions = open.filter(u => u[0] === 'region');
    for (const [lineKind, axis] of [['row', 0], ['col', 1]]) {
      const openLines = open.filter(u => u[0] === lineKind);
      for (const combo of combos(openRegions, k)) {
        const cells = combo.flatMap(u => st.candidates(u));
        const lines = [...new Set(cells.map(c => c[axis]))].sort((a, b) => a - b);
        if (lines.length !== k) continue;
        const regionSet = new Set(combo.map(u => u[1]));
        const targets = lines.flatMap(line => st.candidates([lineKind, line]).filter(([r, c]) => !regionSet.has(p.regions[r][c])));
        if (targets.length) return { kind: 'eliminate', cells: targets, reason: `${joinNames(combo.map(p.unitName))} ${k === 1 ? 'fits' : 'fit'} only in ${joinNames(lines.map(l => p.unitName([lineKind, l])))}, so the other cells there are out`, cause: cells };
      }
      for (const combo of combos(openLines, k)) {
        const cells = combo.flatMap(u => st.candidates(u));
        const regs = [...new Set(cells.map(([r, c]) => p.regions[r][c]))].sort((a, b) => p.labels.indexOf(a) - p.labels.indexOf(b));
        if (regs.length !== k) continue;
        const lineSet = new Set(combo.map(u => u[1]));
        const targets = regs.flatMap(l => st.candidates(['region', l]).filter(c => !lineSet.has(c[axis])));
        if (targets.length) { const where = joinNames(regs.map(l => p.unitName(['region', l]))); return { kind: 'eliminate', cells: targets, reason: `${joinNames(combo.map(p.unitName))} only ${k === 1 ? 'has' : 'have'} cells of ${where}, so those queens sit there and the rest of ${where} is out`, cause: cells }; }
      }
    }
    return null;
  }
  function qBlocked(st) {
    const p = st.p, open = st.openUnits(), byUnit = [];
    for (let r = 0; r < p.n; r++) for (let c = 0; c < p.n; c++) {
      if (!st.cand[r][c]) continue;
      const struck = st.struckBy([r, c]), own = p.cellUnits([r, c]);
      for (const u of open) {
        if (own.some(o => sameUnit(o, u))) continue;
        if (st.candidates(u).every(([rr, cc]) => struck.has(rr * p.n + cc))) {
          let entry = byUnit.find(e => sameUnit(e.unit, u));
          if (!entry) { entry = { unit: u, cells: [] }; byUnit.push(entry); }
          entry.cells.push([r, c]); break;
        }
      }
    }
    if (!byUnit.length) return null;
    let best = byUnit[0]; for (const e of byUnit) if (e.cells.length > best.cells.length) best = e;
    return { kind: 'eliminate', cells: best.cells, reason: `a queen on ${best.cells.length > 1 ? 'any of these cells' : 'this cell'} would leave ${p.unitName(best.unit)} with no room`, cause: st.candidates(best.unit) };
  }
  function qApply(st, step) { step.removed = step.kind === 'queen' ? st.place(step.cells[0]) : st.eliminate(step.cells); return step; }
  function qNextStep(st) {
    let step = qSingle(st);
    if (step === 'contradiction') return step;
    step = step || qConfine(st, 1) || qBlocked(st);
    for (let k = 2; k < st.p.n && !step; k++) step = qConfine(st, k);
    return step;
  }
  function qDeduce(p) {
    const st = new QState(p), steps = [];
    while (st.queens.length < p.n) {
      const step = qNextStep(st);
      if (!step || step === 'contradiction') break;
      qApply(st, step);
      if (!st.ok()) break;
      steps.push(step);
    }
    let decided = 0; for (const row of st.cand) for (const v of row) if (!v) decided++;
    return { steps, done: st.queens.length === p.n, progress: decided + 3 * st.queens.length, queens: st.queens };
  }
  return { QPALETTE, queensPuzzle, QState, qNextStep, qApply, qDeduce };
}

File: queens/solver.py
"""Deterministic solver for LinkedIn Queens puzzles.

Rules: place one queen in every row, every column and every colour region;
no two queens may touch, diagonally included.

Puzzle text format: N lines of N region labels (letters or digits).
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import dataclass, field
from itertools import combinations
from pathlib import Path

Coord = tuple[int, int]
Unit = tuple[str, int | str]

PALETTE = [
    ("lavender", "#BBA3E2"), ("blue", "#96BEFF"), ("orange", "#FFC992"), ("gray", "#DFDFDF"),
    ("green", "#B3DFA0"), ("taupe", "#B9B29E"), ("red", "#FF7B60"), ("yellow", "#E6F388"),
    ("pink", "#DFA0BF"), ("teal", "#95CBCF"), ("salmon", "#FAA889"), ("cyan", "#55EBE2"),
    ("magenta", "#FE93F1"), ("mint", "#91F5AD"), ("periwinkle", "#C9C9EE"), ("emerald", "#5BBA6F"),
]
HEX_NAMES = {hex_: name for name, hex_ in PALETTE} | {
    "#D9D9D9": "gray", "#D895B2": "pink", "#F96C51": "red", "#AF96DC": "purple", "#A6D995": "green",
    "#FBBF81": "orange", "#85B5FC": "blue", "#DCF079": "lime", "#ADA68E": "taupe", "#FFFFFF": "white",
}


class Contradiction(Exception):
    pass


class Unsolvable(Exception):
    pass


@dataclass
class Puzzle:
    n: int
    regions: list[list[str]]
    colors: dict[str, str] = field(default_factory=dict)
    source: str = ""
    labels: list[str] = field(init=False)
    region_cells: dict[str, list[Coord]] = field(init=False)
    names: dict[str, str] = field(init=False)

    def __post_init__(self) -> None:
        self.labels = []
        self.region_cells = {}
        for r in range(self.n):
            for c in range(self.n):
                label = self.regions[r][c]
                if label not in self.region_cells:
                    self.labels.append(label)
                    self.region_cells[label] = []
                self.region_cells[label].append((r, c))
        if len(self.labels) != self.n:
            raise ValueError(f"{self.n}x{self.n} grid needs {self.n} regions, found {len(self.labels)}")
        for i, label in enumerate(self.labels):
            self.colors.setdefault(label, PALETTE[i % len(PALETTE)][1])
        self.names = {}
        seen: dict[str, int] = {}
        for label in self.labels:
            name = HEX_NAMES.get(self.colors[label].upper(), "colour " + label)
            seen[name] = seen.get(name, 0) + 1
            self.names[label] = name
        for label in self.labels:
            if seen[self.names[label]] > 1:
                self.names[label] += f" ({label})"

    @classmethod
    def parse(cls, text: str) -> Puzzle:
        colors: dict[str, str] = {}
        source = ""
        rows = []
        for line in text.splitlines():
            line = line.strip()
            if not line:
                continue
            if line.startswith("#"):
                body = line[1:].strip()
                if body.startswith("colors:"):
                    for item in body[len("colors:"):].split():
                        label, _, hex_ = item.partition("=")
                        colors[label] = hex_
                elif body.startswith("source:"):
                    source = body[len("source:"):].strip()
                continue
            rows.append(line)
        n = len(rows)
        if any(len(row) != n for row in rows):
            raise ValueError("grid must be square")
        return cls(n, [list(row) for row in rows], colors, source)

    def render(self) -> str:
        lines = []
        if self.source:
            lines.append(f"# source: {self.source}")
        lines.append("# colors: " + " ".join(f"{label}={self.colors[label]}" for label in self.labels))
        lines += ["".join(row) for row in self.regions]
        return "\n".join(lines)

    def color_name(self, label: str) -> str:
        return self.names[label]

    def units(self) -> list[Unit]:
        return (
            [("row", i) for i in range(self.n)]
            + [("col", i) for i in range(self.n)]
            + [("region", label) for label in self.labels]
        )

    def unit_cells(self, unit: Unit) -> list[Coord]:
        kind, key = unit
        if kind == "row":
            return [(key, c) for c in range(self.n)]  # type: ignore[misc]
        if kind == "col":
            return [(r, key) for r in range(self.n)]  # type: ignore[misc]
        return self.region_cells[key]  # type: ignore[index]

    def unit_name(self, unit: Unit) -> str:
        kind, key = unit
        if kind == "row":
            return f"row {key + 1}"  # type: ignore[operator]
        if kind == "col":
            return f"column {key + 1}"  # type: ignore[operator]
        return f"the {self.color_name(key)} region"  # type: ignore[arg-type]

    def cell_units(self, cell: Coord) -> list[Unit]:
        r, c = cell
        return [("row", r), ("col", c), ("region", self.regions[r][c])]

    def neighbours(self, cell: Coord) -> list[Coord]:
        r, c = cell
        return [
            (r + dr, c + dc)
            for dr in (-1, 0, 1)
            for dc in (-1, 0, 1)
            if (dr or dc) and 0 <= r + dr < self.n and 0 <= c + dc < self.n
        ]


@dataclass
class Step:
    kind: str
    cells: list[Coord]
    reason: str
    cause: list[Coord]
    removed: list[Coord] = field(default_factory=list)

    def __str__(self) -> str:
        where = ", ".join(f"({r + 1},{c + 1})" for r, c in self.cells)
        return f"{self.kind} {where}  [{self.reason}]"


class State:
    def __init__(self, puzzle: Puzzle) -> None:
        self.p = puzzle
        self.cand = [[True] * puzzle.n for _ in range(puzzle.n)]
        self.queens: set[Coord] = set()

    def copy(self) -> State:
        s = State(self.p)
        s.cand = [row[:] for row in self.cand]
        s.queens = set(self.queens)
        return s

    def candidates(self, unit: Unit) -> list[Coord]:
        return [cell for cell in self.p.unit_cells(unit) if self.cand[cell[0]][cell[1]]]

    def has_queen(self, unit: Unit) -> bool:
        return any(q in self.p.unit_cells(unit) for q in self.queens)

    def open_units(self) -> list[Unit]:
        return [u for u in self.p.units() if not self.has_queen(u)]

    def eliminate(self, cells: list[Coord]) -> list[Coord]:
        removed = []
        for r, c in cells:
            if self.cand[r][c]:
                self.cand[r][c] = False
                removed.append((r, c))
        return removed

    def struck_by(self, cell: Coord) -> set[Coord]:
        """Cells a queen on `cell` rules out."""
        out: set[Coord] = set()
        for unit in self.p.cell_units(cell):
            out.update(self.p.unit_cells(unit))
        out.update(self.p.neighbours(cell))
        out.discard(cell)
        return out

    def place(self, cell: Coord) -> list[Coord]:
        self.queens.add(cell)
        self.cand[cell[0]][cell[1]] = False
        return self.eliminate(sorted(self.struck_by(cell)))

    def check(self) -> None:
        for unit in self.open_units():
            if not self.candidates(unit):
                raise Contradiction(f"{self.p.unit_name(unit)} has no cell left")


def find_single(state: State) -> Step | None:
    p = state.p
    for unit in state.open_units():
        cands = state.candidates(unit)
        if not cands:
            raise Contradiction(f"{p.unit_name(unit)} has no cell left")
        if len(cands) == 1:
            cell = cands[0]
            cause = [c for c in p.unit_cells(unit) if c != cell]
            return Step("queen", [cell], f"{p.unit_name(unit)} has one cell left", cause)
    return None


def find_confine(state: State, k: int) -> Step | None:
    p = state.p
    open_regions = [u for u in state.open_units() if u[0] == "region"]
    for line_kind, axis in (("row", 0), ("col", 1)):
        open_lines = [u for u in state.open_units() if u[0] == line_kind]
        for combo in combinations(open_regions, k):
            cells = [c for u in combo for c in state.candidates(u)]
            lines = {c[axis] for c in cells}
            if len(lines) != k:
                continue
            region_set = {u[1] for u in combo}
            targets = [
                c for line in sorted(lines) for c in state.candidates((line_kind, line))
                if p.regions[c[0]][c[1]] not in region_set
            ]
            if targets:
                names = join_names([p.unit_name(u) for u in combo])
                where = join_names([p.unit_name((line_kind, line)) for line in sorted(lines)])
                verb = "fits" if k == 1 else "fit"
                return Step("eliminate", targets, f"{names} {verb} only in {where}, so the other cells there are out", cells)
        for combo in combinations(open_lines, k):
            cells = [c for u in combo for c in state.candidates(u)]
            regions = {p.regions[r][c] for r, c in cells}
            if len(regions) != k:
                continue
            line_set = {u[1] for u in combo}
            targets = [
                c for label in sorted(regions, key=p.labels.index) for c in state.candidates(("region", label))
                if c[axis] not in line_set
            ]
            if targets:
                names = join_names([p.unit_name(u) for u in combo])
                where = join_names([p.unit_name(("region", label)) for label in sorted(regions, key=p.labels.index)])
                has = "has" if k == 1 else "have"
                return Step("eliminate", targets, f"{names} only {has} cells of {where}, so those queens sit there and the rest of {where} is out", cells)
    return None


def find_blocked(state: State) -> Step | None:
    p = state.p
    open_units = state.open_units()
    by_unit: dict[Unit, list[Coord]] = {}
    for r in range(p.n):
        for c in range(p.n):
            if not state.cand[r][c]:
                continue
            struck = state.struck_by((r, c))
            own = set(p.cell_units((r, c)))
            for unit in open_units:
                if unit in own:
                    continue
                if all(cell in struck for cell in state.candidates(unit)):
                    by_unit.setdefault(unit, []).append((r, c))
                    break
    if not by_unit:
        return None
    unit, cells = max(by_unit.items(), key=lambda kv: len(kv[1]))
    plural = "any of these cells" if len(cells) > 1 else "this cell"
    return Step("eliminate", cells, f"a queen on {plural} would leave {p.unit_name(unit)} with no room", state.candidates(unit))


def basic_propagate(state: State) -> None:
    while True:
        step = find_single(state) or find_confine(state, 1) or find_blocked(state)
        if step is None:
            return
        apply_step(state, step)
        state.check()


def find_by_contradiction(state: State) -> Step | None:
    p = state.p
    for r in range(p.n):
        for c in range(p.n):
            if not state.cand[r][c]:
                continue
            trial = state.copy()
            trial.place((r, c))
            try:
                trial.check()
                basic_propagate(trial)
            except Contradiction as exc:
                return Step("eliminate", [(r, c)], f"a queen at ({r + 1},{c + 1}) leads to a dead end: {exc}", [])
    return None


def apply_step(state: State, step: Step) -> None:
    if step.kind == "queen":
        step.removed = state.place(step.cells[0])
    else:
        step.removed = state.eliminate(step.cells)


def solve(puzzle: Puzzle) -> tuple[set[Coord], list[Step]]:
    state = State(puzzle)
    steps: list[Step] = []
    while len(state.queens) < puzzle.n:
        step = find_single(state) or find_confine(state, 1) or find_blocked(state)
        for k in range(2, puzzle.n):
            if step is not None:
                break
            step = find_confine(state, k)
        if step is None:
            step = find_by_contradiction(state)
        if step is None:
            raise Unsolvable("no further deduction possible")
        apply_step(state, step)
        state.check()
        steps.append(step)
    check_solution(puzzle, state.queens)
    return state.queens, steps


def check_solution(puzzle: Puzzle, queens: set[Coord]) -> None:
    if len(queens) != puzzle.n:
        raise Unsolvable(f"{len(queens)} queens placed, need {puzzle.n}")
    for unit in puzzle.units():
        if sum(q in puzzle.unit_cells(unit) for q in queens) != 1:
            raise Unsolvable(f"{puzzle.unit_name(unit)} does not have exactly one queen")
    for q in queens:
        if any(nb in queens for nb in puzzle.neighbours(q)):
            raise Unsolvable(f"queen at {q} touches another queen")


def join_names(names: list[str]) -> str:
    if len(names) <= 1:
        return names[0] if names else ""
    return ", ".join(names[:-1]) + " and " + names[-1]


def trace(puzzle: Puzzle, steps: list[Step], title: str) -> dict:
    return {
        "title": title,
        "n": puzzle.n,
        "regions": ["".join(row) for row in puzzle.regions],
        "labels": puzzle.labels,
        "colors": [puzzle.colors[label] for label in puzzle.labels],
        "names": [puzzle.color_name(label) for label in puzzle.labels],
        "source": puzzle.source,
        "steps": [
            {"kind": s.kind, "cells": s.cells, "reason": s.reason, "cause": s.cause, "removed": s.removed}
            for s in steps
        ],
    }


def write_html(puzzle: Puzzle, steps: list[Step], title: str, out: Path) -> None:
    """Write a single-puzzle page using the shared viewer template."""
    assemble = _load_assembler()
    queens, _ = solve(puzzle)
    entry = {
        "game": "queens", "id": title, "title": title, "group": "Puzzle", "source": puzzle.source, "n": puzzle.n,
        "regions": ["".join(row) for row in puzzle.regions],
        "labels": puzzle.labels,
        "colors": [puzzle.colors[label] for label in puzzle.labels],
        "names": [puzzle.color_name(label) for label in puzzle.labels],
        "solution": sorted(queens),
    }
    out.write_text(assemble.inline({"puzzles": [entry]}))


def _load_assembler():
    import importlib.util

    path = Path(__file__).parent.parent / "web" / "assemble.py"
    spec = importlib.util.spec_from_file_location("web_assemble", path)
    module = importlib.util.module_from_spec(spec)
    assert spec.loader is not None
    spec.loader.exec_module(module)
    return module


def render_solution(puzzle: Puzzle, queens: set[Coord]) -> str:
    return "\n".join(
        "".join("Q" if (r, c) in queens else puzzle.regions[r][c].lower() for c in range(puzzle.n))
        for r in range(puzzle.n)
    )


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Solve a Queens puzzle by deduction.")
    parser.add_argument("puzzle", type=Path)
    parser.add_argument("-v", "--verbose", action="store_true", help="print each deduction step")
    parser.add_argument("--html", type=Path, metavar="OUT", help="write a playable board and step viewer")
    args = parser.parse_args(argv)

    puzzle = Puzzle.parse(args.puzzle.read_text())
    try:
        queens, steps = solve(puzzle)
    except (Contradiction, Unsolvable) as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1
    if args.verbose:
        for step in steps:
            print(step)
        print()
    print(render_solution(puzzle, queens))
    if args.html:
        write_html(puzzle, steps, args.puzzle.stem, args.html)
        print(f"wrote {args.html}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
